ssl/statem/statem_srvr.c

```diff
@@ -2592,21 +2586,21 @@ int tls_construct_server_key_exchange(SSL *s, WPACKET *pkt)
             goto err;
         }
 
-        dh = EVP_PKEY_get0_DH(s->s3.tmp.pkey);
-        if (dh == NULL) {
-            SSLfatal(s, SSL_AD_INTERNAL_ERROR,
-                     SSL_F_TLS_CONSTRUCT_SERVER_KEY_EXCHANGE,
-                     ERR_R_INTERNAL_ERROR);
-            goto err;
-        }
-
         EVP_PKEY_free(pkdh);
         pkdh = NULL;
 
-        DH_get0_pqg(dh, &r[0], NULL, &r[1]);
-        DH_get0_key(dh, &r[2], NULL);
+        /* These BIGNUMs need to be freed when we're finished */
+        freer = 1;
```

Comment:
Instead of using this freer flag initialize the r array with NULLs. This is incorrect because theoretically any of the EVP_PKEY_get_bn_param could fail leaving the rest of the r uninitialized.

Reply:
Err, they are already initialised to NULL earlier on in this function. We still need to use the `freer` variables, because other codepaths populate the r array with BIGNUMs that should *no*t be freed. However, this one populates it with values that *should* be freed.

Reply:
Ah, I overlooked the initialization in the unchanged code.